### ISSUE TITLE
not enough funds

### ISSUE THREAD

**Issue by nicbus** [BP-WG/bp-wallet#13]
Re-opening here, as indicated in https://github.com/RGB-WG/rgb/issues/113

Trying to execute a few transfers using this [branch](https://github.com/nicbus/rgb-sandbox/tree/not_enough_funds) of rgb-sandbox I run into the following error:
```
Error: not enough funds to pay fee of 400 sats; all inputs contain 2000 sats and outputs spends 2000 sats out of them.
```

The wallet should have enough bitcoins as the unspents come up as:
```
Height     Amount, ṩ    Outpoint                                                                                                                                                                                                                                                                                               
bcrt1q06vjlq89kufjgzenwusm99l6hv5ygxhur8qun8    &9/0                                                                                                                                                                                                                                                                           
107             2000    e6cfb776a1e03165f6e739a6452231ac9d30fb20cbde0ae130ed7b10d20dc7b1:0                                                                                                                                                                                                                                     
106             2000    7a39dd007218bdbf3d6d1a2b919700d9a144a1d2cab20cace5a8ad57df2395b7:0                                                                                                                                                                                                                                     
                                                                                                                                                                                                                                                                                                                               
bcrt1qwke7jgvxuwcwk8q4m4mglgmltyfzasde5c4nsj    &0/0                                                                                                                                                                                                                                                                           
105        100000000    2a0bd1672cfb051c7c0627d1295c07325974b53048a87169c80b44f8698c949e:0                                                                                                                                                                                                                                     
```

I guess this happens because the input selection is not adding more inputs if needed to cover fees (or outputs), but I haven't dug deeper than this.

**Comment by dr-orlovsky** [comment]
@nicbus can you please test #17 for the solution?

**Comment by nicbus** [comment]
Tested it using this [updated](https://github.com/nicbus/rgb-sandbox/tree/not_enough_funds_136) rgb-sandbox branch.

Compared to the original branch it updates to `rgb-wallet` from the `v0.11` branch, with `bp-wallet` patched from branch `fee` (from #17).

I also updated the sandbox run to better stimulate the issue, it now:
- executes a 1st transfer, wallet 1 to wallet 2, using witness
- adds funds to wallet 2 on keychain 0
- executes a 2nd transfer, wallet 2 to wallet 3, using witness

With this setup I still see the issue.

**Comment by nicbus** [comment]
Tested with an updated rgb-sandbox [branch](https://github.com/nicbus/rgb-sandbox/tree/not_enough_funds.2024-03-21) that uses the current `master` branch of `rgb-wallet` and the issue is still there.

**Comment by dr-orlovsky** [comment]
I have updated #17 on top of master with all past changes. Can you pls test it once again and provide me with the exact message it generates like in the original issue (all outputs etc)?

I do not see why it may not be working...

**Comment by nicbus** [comment]
Tested it again (rgb-sandbox [branch](https://github.com/nicbus/rgb-sandbox/tree/not_enough_funds.2024-03-26)) using the updated `rgb` branch `dups` (commit `ea858d0`) and what I get when calling `rgb transfer` is:
```
RGB: command-line wallet for RGB smart contracts
     by LNP/BP Standards Association

Loading descriptor from wallet rcpt1 ... success
Loading stock ... success
Error: not enough funds to pay fee of 400 sats; all inputs contain 2000 sats and outputs spends 2000 sats out of them.
```

These are the unspents before the transfer:
```
Height     Amount, ṩ    Outpoint                                                                                                                                                                                                                                                                                               
bcrt1q09vq53cr3cz9sv20gmuqrdncvqzd6vawcph770    &9/0                                                                                                                                                                                                                                                                           
105             2000    8da0fb64132e5d703532f4bca67261c47b03901620a04e0286054d7485c8c89b:0                                                                                                                                                                                                                                     
                                                                                                                                                                                                                                                                                                                               
bcrt1q23r9p0uf2647utlqwaq3zst5srw34ycw5d72rr    &0/0                                                                                                                                                                                                                                                                           
106        100000000    3b627b4638ae702c624b7a1529d32570fa91e2d756f40c826a652f747390a542:1                                                                                                                                                                                                                                     
                                                                                                                                                                                                                                                                                                                               
                                                                                                                                                                                                                                                                                                                               
Wallet total balance: 100002000 ṩ                                                                                                                                                                                                                                                                                              
```

**Comment by dr-orlovsky** [comment]
Alright, I finally get it: there were two issues here:
- we didn't account for the fees in doing coinselect in BP wallet
- we do not account for non-RGB outputs in doing coinselect in RGB wallet

 The first one was (most probably) fixed by #17. The second one is the one which remains.

I am re-opening #113 changing its description to indicate the second issue. I am closing this one and merging #17 since I believe the first issue is fixed. Let me know if you can't do a transfer for 4000 sats when _two_ 3000 sat inputs are present (it should work, but just if not we will re-open this issue)
